OAuth로 네이버 로그인

he2joojo opened the pull request:
## Issues
- #7 

## What is this PR? 👓
OAuth로 회원가입/로그인 기능개발에 대한 PR입니다.

## Key changes 🔑
* 로그인 - 로그인 시 ```ResponseEntity<ApiResponse<LoginResponse>>```을 반환하게 했습니다. 
* 회원가입

## To reviewers 👋
- ```MemberService```에서 커스텀 예외를 던져야 하는 부분에 todo로 명시만 해놨습니다! 브루니 pr 머지하고 추가할게요!
- 로그인 시 ```MemberService```에 Jwt토큰을 반환하는 부분을 추가해야 합니다!
- 회원가입 시 프로필 사진 저장 로직을 추가해야합니다!
- 회원가입 시 동네를 저장해야 하는데 jpa를 사용하면 멤버저장하면서 한번에 가능한지 따로 메서드를 만들어야하는지 모르겠어요!🥹
- 네이버에서 프로필 사진을 가져와서 db에 넣었는데 노션에는 ```이미지 있으면 업로드, 없으면 default 이미지```라고 적혀있어서 수정해야 할 것 같아요~!

he2joojo committed:
#7 feat: 네이버에서 사용자 정보 가져오는 기능 개발

- OAuth를 사용해 네이버에서 사용자의 이메일과 프로필 사진을 가져오는 기능 개발

he2joojo committed:
#7 fix: OauthProperty에 값이 매핑되지 않는 버그 해결

- application-oauth.yml에 naver 제거

he2joojo committed:
#7 feat: 로그인 기능 개발

- JWT토큰 생성 후 반환 시 추가해야합니다.

he2joojo committed:
#7 feat: 회원가입 기능 개발

- loginId가 중복되는지 검증 후 회원을 저장합니다.

he2joojo committed:
#7 feat: 회원가입 기능 개발

- loginId가 중복되는지 검증 후 회원을 저장합니다.

he2joojo committed:
#7 style: todo 주석 추가

23Yong commented on build.gradle:
```diff
@@ -35,6 +35,13 @@ dependencies {
     testImplementation 'org.springframework.boot:spring-boot-starter-test'
+
+    // webflux
+    implementation 'org.springframework.boot:spring-boot-starter-webflux'
```
webflux dependency 를 추가하셨군요!

저도 몰랐었는데 `RestTemplate`을 사용해도 괜찮겠더라구요
https://www.youtube.com/watch?v=S4W3cJOuLrU

he2joojo replied:
```webflux```가 무거운 라이브러리인지 몰랐는데 영상보고 처음 알았어요! 수정했습니다:)

23Yong commented on src/main/java/kr/codesquad/secondhand/application/MemberService.java:
```diff
@@ -0,0 +1,114 @@
+
+@RequiredArgsConstructor
+@Service
+public class MemberService {
```
class 레벨에 `@Transactional(readOnly = true)`를 달고 트랜잭션이 필요한 메서드에서는 `@Transactional` 을 다는 건 어떠신가요?

지금은 서비스레이어에 `@Transactional`을 통한 트랜잭션이 걸려있지 않네요!

23Yong commented on src/main/java/kr/codesquad/secondhand/application/MemberService.java:
```diff
@@ -0,0 +1,114 @@
+
+    public LoginResponse login(LoginRequest request, String code) {
+        OauthTokenResponse tokenResponse = getToken(code);
+        UserProfile userProfile = getUserProfile(tokenResponse);
```
서비스 레이어에서 `WebClient`를 통해 네이버 인증서버에 요청을 하고 있어요. 
이것보다는 `NaverClient` 혹은 `NaverRequester`같은 클래스를 따로 만들어 이를 빈으로 등록하고 `NaverRequester`에게 요청을 위임하는 건 어떻게 생각하시나요? 그러면 `NaverRequester`가 `code`를 받아서 액세스 토큰을 요청하고 유저 정보를 요청할 수 있을 것 같아요.

ex. UserProfile userProfile = naverRequester.getUserProfile(code);

he2joojo replied:
```WebClient```대신 ```RestTemplate```사용하도록 수정했습니다!

23Yong replied:
혹시 한 번 더 분리하면 어떨까요? 서비스 클래스의 크기가 큰 것 같아서요!
약간 이런식으로? 똑같이 할 필요는 없숩니다!
```java
@Component
public class NaverRequester {

  private final RestTemplate restTemplate;
  private final OauthProvider oAuthProvider;

  public UserProfile getUserProfile(String code) {
    String accessToken = requestAccessToken(code);
    return requestUserProfile(accessToken);
  }

  // requestAccessToken

  // requestUserProfile
}
```
```java
public LoginResponse login(LoginRequest request, String code) {
  naverRequest.getUserProfile(code);
}
```

he2joojo replied:
```getToken```메서드랑 ```getUserProfile```메서드를 분리하니까 훨씬 깔끔하네요! 👍

23Yong commented on src/main/java/kr/codesquad/secondhand/application/MemberService.java:
```diff
@@ -0,0 +1,114 @@
+    public void signUp(SignUpRequest request, String code) {
+        OauthTokenResponse tokenResponse = getToken(code);
+        UserProfile userProfile = getUserProfile(tokenResponse);
+        verifyDuplicated(request);
```
개인적으로 먼저 중복을 검사하면 좋을 것 같아요.
중복된 아이디면 아예 회원가입 조차 안되니까 Naver OAuth 서버랑 왔다갔다 할 필요가 없을 것 같아서요!

he2joojo replied:
헉 맞네요! 수정하겠습니다:)

23Yong commented on src/main/java/kr/codesquad/secondhand/application/MemberService.java:
```diff
@@ -0,0 +1,114 @@
+        Optional<Member> existingMember = memberRepository.findByLoginId(request.getLoginId());
+        if (existingMember.isPresent()) {
+            throw new IllegalArgumentException(); // todo: 예외 던지기(존재하는 loginId)
+        }
```
```java
if (memberRepository.existsByLoginId(request.getLoginId())) {
  throw new ~
}
```
이렇게 `exists`도 가능합니다!

he2joojo replied:
오오 jpa가 이런 메서드도 만들어줄 수 있는지 몰랐어요! 역시 jpa 고수 브루니..

23Yong commented on src/main/java/kr/codesquad/secondhand/application/MemberService.java:
```diff
@@ -0,0 +1,114 @@
+                .bodyValue(tokenRequest(code))
+                .retrieve()
+                .bodyToMono(OauthTokenResponse.class)
+                .block();
```
```java
.blockOptional()
.orElseThrow();
```
이렇게 하는 것도 좋아하는 편입니다. 그런데 굳이 바꾸실 필요는 없어여!

23Yong commented on src/main/java/kr/codesquad/secondhand/application/MemberService.java:
```diff
@@ -0,0 +1,114 @@
+    }
+
+    private Map<String, Object> getUserAttributes(OauthTokenResponse tokenResponse) {
+        return WebClient.create()
```
`WebClient`가 반복적으로 사용되고 있네용. 빈으로 등록해서 사용하는건 어떤가요?

23Yong commented on src/main/java/kr/codesquad/secondhand/config/OauthConfig.java:
```diff
@@ -0,0 +1,22 @@
+public class OauthConfig {
+
+    private final OauthProperty property;
+
```
빈 라인 하나 지워주세요~ㅎㅎ

23Yong commented on src/main/java/kr/codesquad/secondhand/infrastructure/OauthProperty.java:
```diff
@@ -0,0 +1,31 @@
+import org.springframework.boot.context.properties.ConfigurationProperties;
+
+@Getter
+@ConfigurationProperties(prefix = "oauth2")
```
`@ConfigurationProperties` 👍

23Yong commented on src/main/java/kr/codesquad/secondhand/infrastructure/OauthProperty.java:
```diff
@@ -0,0 +1,31 @@
+    private final Provider provider = new Provider();
+
+    @Getter
+    @Setter
```
`@Setter`는 왜 사용하셨나요?

he2joojo replied:
```@Setter```가 없으면 값을 매핑할때 에러가 나는 것 같아요~!
> Failed to bind properties under 'oauth2.provider' to kr.codesquad.secondhand.infrastructure.OauthProperties$Provider:
Reason: java.lang.IllegalStateException: No setter found for property: token-url

he2joojo replied:
아래있는 브루니 말대로 변수를 ```final```로 수정하고 ```@RequiredArgsConstructor```사용하니까 ```@Setter```생략해도 문제없네요! 감사합니다~!

23Yong commented on src/main/java/kr/codesquad/secondhand/presentation/dto/LoginRequest.java:
```diff
@@ -0,0 +1,19 @@
+public class LoginRequest {
+
+    @NotBlank
+    @Size(min = 2, max = 12) // todo: ExceptionHandler에서 예외처리하기
```
validation의 사용이 좋네용
`@Size(min = 2, max = 12, message = "")`와 같이 메시지도 같이 남겨주면 어떨까요?

23Yong commented on src/main/java/kr/codesquad/secondhand/presentation/dto/SignUpRequest.java:
```diff
@@ -0,0 +1,22 @@
+
+    @NotBlank
+    @Size(min = 2, max = 12)
+    private String loginId;
```
여기도 `message` attribute를 통해 메시지를 남기면 좋을 것 같아용

23Yong commented on src/main/java/kr/codesquad/secondhand/infrastructure/OauthProperty.java:
```diff
@@ -0,0 +1,31 @@
+
+    private final User user = new User();
+    private final Provider provider = new Provider();
+
```
이렇게 `new`연산자를 통해서도 할 수 있지만 `@ConstructorBinding` 을 사용해서도 가능해요!
```java
  private final User user;
  private final Provider provider;

  @ConstructorBinding
  public OauthProperties(User user, Provider provider) {
    this.user = user;
    this.provider = provider;
  }
```

he2joojo replied:
우와! 첨봐요 좋은 어노테이션 많네요!ㅎㅎ

23Yong commented on src/main/java/kr/codesquad/secondhand/infrastructure/OauthProperty.java:
```diff
@@ -0,0 +1,31 @@
+
+        private String tokenUrl;
+        private String userInfoUrl;
+        private String userNameAttribute;
```
`private final`로 모두 수정하고 `@RequiredArgsConstructor`를 다는 건 어떻게 생각하세요?

23Yong commented on src/main/resources/application.yml:
```diff
@@ -2,3 +2,5 @@ spring:
     active:
       - local
+    include:
+      - oauth
```
`include` 👍

23Yong commented on .gitignore:
```diff
@@ -32,3 +32,4 @@ replay_pid*
 
 # GRADLE
 .gradle
+/src/main/resources/application-oauth.yml
```
요렇게 해도 되지만 아래처럼 그냥 경로를 안적고 할 수도 있어요~
```
application-oauth.yml
```

23Yong reviewed (commented):
늦은시간까지 수고하셨습니다! 👍

리뷰 천천히보시구 천천히 수정해주세요~

he2joojo committed:
#7 refactor: WebClient대신 RestTemplate을 사용하도록 수정

he2joojo committed:
#7 refactor: 리뷰 반영 수정

- requestDto 검증 시 message 추가
- 중복아이디 검증시 existsByLoginId 사용하도록 수정
- 로그인, 회원가입시 @Transactional 추가
- oauth.yml 매핑시 @ConstructorBinding 사용하도록 수정

23Yong commented on src/main/java/kr/codesquad/secondhand/application/MemberService.java:
```diff
@@ -21,18 +21,21 @@
     private final MemberRepository memberRepository;
     private final RestTemplate restTemplate;
 
+    @Transactional
```
지금은 DB에 쓰는 동작이 없는 것 같아요! `@Transactional(readOnly = true)`라서 제거해도 될 것 같습니다!

23Yong reviewed (commented):
분리하면 진짜 끝..!일 것 같습니다ㅎ..!

he2joojo committed:
#7 refactor: 리뷰 반영 추가 수정

- 로그인, 회원가입시 @Transactional 제거
- getToken(), getUserProfile()을 NaverRequester 클래스로 분리

23Yong reviewed (approved):
리뷰 반영하느라 수고많으셨습니다! 머지 해도 될 것 같아요!
예외처리는 다음 PR에서 하시죠!

23Yong committed:
Merge branch 'develop' into feature/#7-oauth
